internal/controller/controlplane/k0s_controlplane_controller.go

```diff
@@ -0,0 +1,237 @@
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	kubeadmbootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util"
+	capiutil "sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	"sigs.k8s.io/cluster-api/util/secret"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	defaultK0SVersion = "v1.27.2-k0s.0"
+)
+
+type K0sController struct {
+	client.Client
+	Scheme     *runtime.Scheme
+	ClientSet  *kubernetes.Clientset
+	RESTConfig *rest.Config
+}
+
+// +kubebuilder:rbac:groups=controlplane.cluster.x-k8s.io,resources=k0scontrolplanes/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=controlplane.cluster.x-k8s.io,resources=k0scontrolplanes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=*,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;clusters/status,verbs=get;list;watch;update;pacth
+
+func (c *K0sController) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	log := log.FromContext(ctx).WithValues("controlplane", req.NamespacedName)
+	log.Info("Reconciling K0sControlPlane")
+
+	kcp := &cpv1beta1.K0sControlPlane{}
+	if err := c.Get(ctx, req.NamespacedName, kcp); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Failed to get K0sControlPlane")
+		return ctrl.Result{}, err
+	}
+
+	if !kcp.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("K0sControlPlane is being deleted, no action needed")
+		return ctrl.Result{}, nil
+	}
+
+	if kcp.Spec.K0sVersion == "" {
+		kcp.Spec.K0sVersion = defaultK0SVersion
+	}
+
+	cluster, err := capiutil.GetOwnerCluster(ctx, c.Client, kcp.ObjectMeta)
+	if err != nil {
+		log.Error(err, "Failed to get owner cluster")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		log.Info("Waiting for Cluster Controller to set OwnerRef on K0sControlPlane")
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("cluster", cluster.Name)
+
+	if annotations.IsPaused(cluster, kcp) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	if err := c.ensureCertificates(ctx, cluster, kcp); err != nil {
+		log.Error(err, "Failed to ensure certificates")
+		return ctrl.Result{}, err
+	}
+
+	res, err = c.reconcile(ctx, cluster, kcp)
+	if err != nil {
+		return res, err
+	}
+
+	// TODO: We need to have bit more detailed status and conditions handling
+	kcp.Status.Ready = true
+	kcp.Status.ExternalManagedControlPlane = true
```

Comment:
AFAIK this should be false when we manage CP in `Machine`s

Reply:
Done